### PR TITLE
Needle rifle. Hopefully now.

### PR DESCRIPTION
**Pull request by mrturkeytoe**
As the title suggests.
As of now the stats are as follow:
25 Damage
2 Accuracy (like the carbine)
21 Capacity
3 Shot super combine

### CODE REVIEW

File: code/modules/halo/weapons/covenant/ammo.dm [hunk 1]
```diff
@@ -158,3 +158,59 @@
...
 	icon_state = "carbine_trail"
+
+/obj/item/ammo_magazine/rifleneedlepack
+	name = "Rifle Needles"
+	desc = "A pack of fewer, larger crystalline needles. For T-31 rifle."
+	icon = 'code/modules/halo/icons/Covenant_Projectiles.dmi'
+	icon_state = "needlerpack"
+	max_ammo = 21
+	ammo_type = /obj/item/ammo_casing/rifleneedle
+	caliber = "cov_carbine"
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/ammo.dm]: This needs a unique caliber

File: code/modules/halo/weapons/covenant/ammo.dm [hunk 2]
```diff
@@ -158,3 +158,59 @@
...
+	mag_type = MAGAZINE
+
+/obj/item/ammo_casing/rifleneedle
+	name = "Rifle Needle"
+	desc = "A large crystalline needle"
+	caliber = "cov_carbine"
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/ammo.dm]: This should match the caliber of the ammo magazine

File: code/modules/halo/weapons/covenant/ammo.dm [hunk 3]
```diff
@@ -158,3 +158,59 @@
...
+	projectile_type = /obj/item/projectile/bullet/covenant/rifleneedle
+	icon = 'code/modules/halo/icons/Covenant_Projectiles.dmi'
+	icon_state = "needle"
+
+/obj/item/projectile/bullet/covenant/rifleneedle
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/ammo.dm]: This should be a subtype of the normal needler projectile to reduce duplication

File: code/modules/halo/weapons/covenant/snipers.dm [hunk 4]
```diff
@@ -76,3 +76,40 @@
...
+	name = "Type-31 Needle Rifle"
+	desc = "A unique combination of the Type-33 and Type-51."
+	icon = 'code/modules/halo/icons/Covenant Weapons.dmi'
+	icon_state = "Needle rifle"
+	item_state = "Needle rifle"
+	slot_flags = SLOT_BACK
+	fire_sound = 'code/modules/halo/sounds/cov_needlerifle_fire.ogg'
+	magazine_type = /obj/item/ammo_magazine/rifleneedlepack
+	handle_casings = CLEAR_CASINGS
+	caliber = "cov_carbine"
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/snipers.dm]: This should have a unique caliber

File: code/modules/halo/weapons/covenant/snipers.dm [hunk 5]
```diff
@@ -76,3 +76,40 @@
...
+		)
+
+/obj/item/weapon/gun/projectile/type31needlerifle/verb/scope()
+	set category = "Weapon"
+	set name = "Use Scope"
+	set popup_menu = 1
+
+	toggle_scope(usr, 1.25)
+
+/obj/item/weapon/gun/projectile/type31needlerifle/load_ammo(var/item/I,var/mob/user)
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/snipers.dm]: This shouldn't be here, it's special handling for the carbine only.

File: code/modules/halo/weapons/covenant/snipers.dm [hunk 6]
```diff
@@ -76,3 +76,40 @@
...
+	unload_ammo(user,1)
+	. = ..()
+
+/obj/item/weapon/gun/projectile/type31needlerifle/unload_ammo(var/mob/user,var/allow_dump = 0)
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/snipers.dm]: Ditto.

File: code/modules/halo/weapons/covenant/ammo.dm [hunk 7]
```diff
@@ -158,3 +158,59 @@
...
+	sharp = 1
+	tracer_type = /obj/effect/projectile/bullet/covenant/rifleneedle
+	tracer_delay_time = 0.5 SECONDS
+	invisibility = 101
+
+/obj/effect/projectile/bullet/covenant/rifleneedle
+	icon = 'code/modules/halo/icons/Covenant_Projectiles.dmi'
+	icon_state = "needlerifle_trail"
+
+obj/item/projectile/bullet/covenant/rifleneedle/attack_mob(var/mob/living/carbon/human/L)
```
Comment (X0-11) [inline code/modules/halo/weapons/covenant/ammo.dm]: This should be removed when the projectile is recoded into a subtype due to that projectile providing variables to customise these parameters in this way.

**Review by X0-11 — CHANGES_REQUESTED** [review]

File: code/modules/halo/weapons/covenant/ammo.dm (hunk 1 above)
Comment (mrturkeytoe) [reply code/modules/halo/weapons/covenant/ammo.dm]: How exactly do Calibers work?

File: code/modules/halo/weapons/covenant/ammo.dm (hunk 3 above)
Comment (mrturkeytoe) [reply code/modules/halo/weapons/covenant/ammo.dm]: And how do projectile subtypes work?

File: code/modules/halo/weapons/covenant/snipers.dm (hunk 5 above)
Comment (mrturkeytoe) [reply code/modules/halo/weapons/covenant/snipers.dm]: You mean the special reload behaviour?

File: code/modules/halo/weapons/covenant/ammo.dm (hunk 1 above)
Comment (caelaislinn) [reply code/modules/halo/weapons/covenant/ammo.dm]: Caliber is just a unique identifier for the ammo type. Rename it to the same thing as the needler ammo.

File: code/modules/halo/weapons/covenant/ammo.dm (hunk 3 above)
Comment (caelaislinn) [reply code/modules/halo/weapons/covenant/ammo.dm]: They inhert procs and variables of the parent, allowing you to override specific members to customise aspects while minimising code duplication. Have a read up on inheritance tutorials in python or c#

Comment (mrturkeytoe) [reply code/modules/halo/weapons/covenant/ammo.dm]: Would it be done in a similar fashion to the relationship between plasma rifle and heavy plasma rifle?

Comment (mrturkeytoe) [reply code/modules/halo/weapons/covenant/ammo.dm]: Also if it's a subtype, wouldn't it display the unwanted homing feature? Or how would one go about making it not do that. When I have it set up as a subtype with the sharp and embed it also seems to have the old issue of not hitting targets that aren't standing up.

**Review by X0-11 — APPROVED** [review]

**Pull request merged**